rust/loqui_client/src/connection_handler.rs

```diff
@@ -0,0 +1,306 @@
+            }
+            DelegatedFrame::Push(_) => None,
+            // XXX: hack for existential types :ablobgrimace:
+            //
+            // The type system needs a concrete impl Future, otherwise it can't infer the `T` of
+            // `Option<T>`. Using None::<Self::HandleFrameFuture> results in a cycle.
+            //
+            // This code path should never be hit if the server implements loqui properly.
+            DelegatedFrame::Request(_) => Some(
+                async {
+                    Err((
+                        LoquiError::InvalidOpcode {
+                            actual: Request::OPCODE,
+                            expected: None,
+                        }
+                        .into(),
+                        0,
+                    ))
+                },
+            ),
+        }
+    }
+
+    fn handle_internal_event(
+        &mut self,
+        event: InternalEvent<E::Encoded, E::Decoded>,
+        id_sequence: &mut IdSequence,
+        transport_options: &TransportOptions,
+    ) -> Option<LoquiFrame> {
+        let TransportOptions { encoding, .. } = transport_options;
+        // Forward Request and Push events to the connection so it can send them to the server.
+        match event {
+            InternalEvent::Request { payload, waiter_tx } => {
+                let sequence_id = id_sequence.next();
+                self.handle_request(payload, encoding, sequence_id, waiter_tx)
+            }
+            InternalEvent::Push { payload } => self.handle_push(payload, encoding),
+        }
+    }
+}
+
+impl<E: Encoder> ConnectionHandler<E> {
+    fn handle_push(&mut self, payload: E::Encoded, encoding: &'static str) -> Option<LoquiFrame> {
+        match self.config.encoder.encode(encoding, payload) {
+            Ok((payload, compressed)) => {
+                let push = Push {
+                    payload,
+                    flags: make_flags(compressed),
+                };
+                Some(push.into())
+            }
+            Err(e) => {
+                error!("Failed to encode payload. error={:?}", e);
+                None
+            }
+        }
+    }
+
+    fn handle_request(
+        &mut self,
+        payload: E::Encoded,
+        encoding: &'static str,
+        sequence_id: u32,
+        waiter_tx: OneShotSender<Result<E::Decoded, Error>>,
+    ) -> Option<LoquiFrame> {
+        match self.config.encoder.encode(encoding, payload) {
+            Ok((payload, compressed)) => {
+                // Store the waiter so we can notify it when we get a response.
+                self.waiters.insert(sequence_id, waiter_tx);
+                let request = Request {
+                    payload,
+                    sequence_id,
+                    flags: make_flags(compressed),
+                };
+                Some(request.into())
+            }
+            Err(e) => {
+                error!("Failed to encode payload. error={:?}", e);
+                None
```

Comment:
This results in a forever-lingering request - maybe we should return this error to the waiter_tx? 

Actually, this may result in a false connection closed error message as the `waiter_tx` will get dropped.